--- a/base/trace_event/builtin_categories.h
+++ b/base/trace_event/builtin_categories.h
@@ -92,6 +92,8 @@ PERFETTO_DEFINE_CATEGORIES_IN_NAMESPACE_
     perfetto::Category("base.power").SetDescription(
       "Events about global system power and battery/thermal state.")
       .SetTags("toplevel"),
+    perfetto::Category("brave"),
+    perfetto::Category("brave.adblock"),
     perfetto::Category("browser").SetTags("navigation"),
     perfetto::Category("browsing_data"),
     perfetto::Category("CacheStorage"),
--- a/chrome/app/chrome_command_ids.h
+++ b/chrome/app/chrome_command_ids.h
@@ -111,6 +111,8 @@
 #define IDC_UNFOCUS_TAB_GROUP         34108
 
 // Page-related commands
+#define IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS 35999
+
 #define IDC_BOOKMARK_THIS_TAB           35000
 #define IDC_BOOKMARK_ALL_TABS           35001
 #define IDC_VIEW_SOURCE                 35002
--- a/chrome/app/generated_resources.grd
+++ b/chrome/app/generated_resources.grd
@@ -578,6 +578,9 @@ are declared in tools/grit/grit_args.gni
 
        <!-- content area context menus. Android does not use it -->
       <if expr="not is_android">
+        <message name="IDS_ADBLOCK_CONTEXT_BLOCK_ELEMENTS" desc="Message for context menu that enables the element picker UI to apply cosmetic filter rules to the page">
+          Block elements
+        </message>
         <message name="IDS_CONTENT_CONTEXT_INSPECTELEMENT" desc="The name of the Inspect Element command in the content area context menu">
           I&amp;nspect
         </message>
--- a/chrome/browser/BUILD.gn
+++ b/chrome/browser/BUILD.gn
@@ -239,6 +239,8 @@ static_library("browser") {
     "command_updater_delegate.h",
     "command_updater_impl.cc",
     "command_updater_impl.h",
+    "cosmetic_filters/cosmetic_filters_tab_helper.cc",
+    "cosmetic_filters/cosmetic_filters_tab_helper.h",
     "defaults.cc",
     "defaults.h",
     "expired_flags_list.h",
@@ -1654,6 +1656,15 @@ static_library("browser") {
     "//chrome/browser/sharing/glic_experimental_triggering",
     "//chrome/browser/sharing/glic_experimental_triggering:impl",
     "//chrome/browser/ui/webauthn:impl",
+
+    # chromium-adblock
+    "//chrome/browser/brave_shields",
+    "//chrome/browser/brave_shields:brave_shields_impl",
+    "//components/brave_shields/core/common",
+    "//components/brave_shields/core/common:mojom",
+    "//components/cosmetic_filters/browser",
+    "//components/cosmetic_filters/common:mojom",
+    "//components/strings:components_strings_grit",
   ]
 
   if (enable_webui_ntp) {
--- a/chrome/browser/browser_process.h
+++ b/chrome/browser/browser_process.h
@@ -92,6 +92,10 @@ namespace activity_reporter {
 class ActivityReporter;
 }
 
+namespace brave_shields {
+class AdBlockService;
+}
+
 namespace component_updater {
 class ComponentUpdateService;
 }
@@ -148,6 +152,8 @@ class BrowserProcess {
 
   virtual ~BrowserProcess();
 
+  virtual brave_shields::AdBlockService* ad_block_service() = 0;
+
   // Returns the UnownedUserDataHost associated with this browser process. This
   // is used to retrieve arbitrary features from the browser process without
   // requiring BrowserProcess to have knowledge of them.
--- a/chrome/browser/browser_process_impl.cc
+++ b/chrome/browser/browser_process_impl.cc
@@ -101,6 +101,13 @@
 #include "chrome/common/url_constants.h"
 #include "chrome/grit/branded_strings.h"
 #include "chrome/installer/util/google_update_settings.h"
+
+// chromium-adblock
+#include "chrome/browser/brave_shields/ad_block_subscription_download_manager_getter.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
+#include "components/brave_shields/content/browser/ad_block_subscription_service_manager.h"
+#include "components/brave_shields/core/common/features.h"
+
 #include "components/activity_reporter/activity_reporter.h"
 #include "components/application_locale_storage/application_locale_storage.h"
 #include "components/breadcrumbs/core/application_breadcrumbs_logger.h"
@@ -725,6 +732,9 @@ void BrowserProcessImpl::PostDestroyThre
 #endif
 
   features_->PostDestroyThreads();
+
+  // AdBlockService should outlive its own worker thread.
+  ad_block_service_.reset();
 }
 #endif  // !BUILDFLAG(IS_ANDROID)
 
@@ -1349,6 +1359,21 @@ StatusTray* BrowserProcessImpl::status_t
   return status_tray_.get();
 }
 
+brave_shields::AdBlockService* BrowserProcessImpl::ad_block_service() {
+  if (!ad_block_service_) {
+    scoped_refptr<base::SequencedTaskRunner> task_runner(
+        base::ThreadPool::CreateSequencedTaskRunner(
+            {base::MayBlock(), base::TaskPriority::USER_BLOCKING,
+             base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN}));
+    ad_block_service_ = std::make_unique<brave_shields::AdBlockService>(
+        local_state(), GetApplicationLocale(), component_updater(), task_runner,
+        AdBlockSubscriptionDownloadManagerGetter(),
+        profile_manager()->user_data_dir().Append(
+            profile_manager()->GetInitialProfileDir()));
+  }
+  return ad_block_service_.get();
+}
+
 #if BUILDFLAG(SAFE_BROWSING_AVAILABLE)
 safe_browsing::SafeBrowsingService*
 BrowserProcessImpl::safe_browsing_service() {
--- a/chrome/browser/browser_process_impl.h
+++ b/chrome/browser/browser_process_impl.h
@@ -515,6 +515,10 @@ class BrowserProcessImpl : public Browse
   std::optional<std::pair<size_t, std::unique_ptr<os_crypt_async::KeyProvider>>>
       additional_provider_for_test_;
 
+  // chromium-adblock
+  brave_shields::AdBlockService* ad_block_service() override;
+  std::unique_ptr<brave_shields::AdBlockService> ad_block_service_;
+
   // Do not add new members to this class. Instead use GlobalFeatures. See
   // browser_process.h file level comments for more details.
 
--- a/chrome/browser/chrome_content_browser_client.cc
+++ b/chrome/browser/chrome_content_browser_client.cc
@@ -764,6 +764,12 @@
 
 #include "base/win/windows_h_disallowed.h"
 
+// chromium-adblock
+#include "chrome/browser/net/brave_proxying_url_loader_factory.h"
+#include "chrome/browser/net/brave_proxying_web_socket.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
+#include "components/brave_shields/content/browser/domain_block_navigation_throttle.h"
+
 using blink::mojom::EffectiveConnectionType;
 using blink::web_pref::WebPreferences;
 using content::BrowserThread;
@@ -5619,6 +5625,21 @@ void ChromeContentBrowserClient::RemoveP
 void ChromeContentBrowserClient::CreateThrottlesForNavigation(
     content::NavigationThrottleRegistry& registry) {
   CreateAndAddChromeThrottlesForNavigation(registry);
+
+  content::NavigationHandle& navigation_handle = registry.GetNavigationHandle();
+  content::BrowserContext* context =
+      navigation_handle.GetWebContents()->GetBrowserContext();
+
+  // The HostContentSettingsMap might be null for some irregular profiles, e.g.
+  // the System Profile.
+  auto* host_content_settings_map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+  if (host_content_settings_map) {
+    brave_shields::DomainBlockNavigationThrottle::MaybeCreateAndAdd(
+        registry, g_browser_process->ad_block_service(),
+        g_browser_process->ad_block_service()->custom_filters_provider(),
+        host_content_settings_map, g_browser_process->GetApplicationLocale());
+  }
 }
 
 std::vector<std::unique_ptr<content::CommitDeferringCondition>>
@@ -6650,6 +6671,9 @@ void ChromeContentBrowserClient::WillCre
     bool* disable_secure_dns,
     network::mojom::URLLoaderFactoryOverridePtr* factory_override,
     scoped_refptr<base::SequencedTaskRunner> navigation_response_task_runner) {
+  BraveProxyingURLLoaderFactory<std::shared_ptr>::MaybeProxyRequest(
+      browser_context, frame, factory_builder, navigation_response_task_runner);
+
 #if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
   auto* web_request_api =
       extensions::BrowserContextKeyedAPIFactory<extensions::WebRequestAPI>::Get(
@@ -6766,24 +6790,7 @@ ChromeContentBrowserClient::
 
 bool ChromeContentBrowserClient::WillInterceptWebSocket(
     content::RenderFrameHost* frame) {
-#if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
-  if (!frame) {
-    return false;
-  }
-  const auto* web_request_api =
-      extensions::BrowserContextKeyedAPIFactory<extensions::WebRequestAPI>::Get(
-          frame->GetBrowserContext());
-
-  // NOTE: Some unit test environments do not initialize
-  // BrowserContextKeyedAPI factories for e.g. WebRequest.
-  if (!web_request_api) {
-    return false;
-  } else {
-    return web_request_api->MayHaveProxiesForFrame(frame);
-  }
-#else
-  return false;
-#endif
+  return (frame != nullptr);
 }
 
 void ChromeContentBrowserClient::CreateWebSocket(
@@ -6794,20 +6801,52 @@ void ChromeContentBrowserClient::CreateW
     const std::optional<std::string>& user_agent,
     mojo::PendingRemote<network::mojom::WebSocketHandshakeClient>
         handshake_client) {
+  // PATCH NOTE: This emulates `BraveContentBrowserClient`, but because
+  // that is essentially a wrapper around this very class, this has to
+  // copy the old ChromeContentBrowserClient::WillInterceptWebSocket
+  // into a lambda here since the logic of that function has been replaced.
+  auto* proxy =
+      BraveProxyingWebSocket<std::shared_ptr>::ProxyWebSocket(
+          frame, std::move(factory), url, site_for_cookies, user_agent);
+
+  auto WillInterceptWebSocketImpl = [](content::RenderFrameHost* frame)-> bool {
 #if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
-  // TODO(crbug.com/40195467): Request w/o a frame also should be proxied.
-  if (!frame) {
-    return;
-  }
-  auto* web_request_api =
-      extensions::BrowserContextKeyedAPIFactory<extensions::WebRequestAPI>::Get(
-          frame->GetBrowserContext());
+    if (!frame) {
+      return false;
+    }
+    const auto* web_request_api =
+        extensions::BrowserContextKeyedAPIFactory<extensions::WebRequestAPI>::Get(
+            frame->GetBrowserContext());
+
+    // NOTE: Some unit test environments do not initialize
+    // BrowserContextKeyedAPI factories for e.g. WebRequest.
+    if (!web_request_api) {
+      return false;
+    } else {
+      return web_request_api->MayHaveProxiesForFrame(frame);
+    }
+#else
+    return false;
+#endif
+  };
 
-  DCHECK(web_request_api);
-  web_request_api->ProxyWebSocket(frame, std::move(factory), url,
-                                  site_for_cookies, user_agent,
-                                  std::move(handshake_client));
+  if (WillInterceptWebSocketImpl(frame)) {
+#if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
+    // TODO(crbug.com/40195467): Request w/o a frame also should be proxied.
+    if (!frame) {
+      return;
+    }
+    auto* web_request_api = extensions::BrowserContextKeyedAPIFactory<
+        extensions::WebRequestAPI>::Get(frame->GetBrowserContext());
+
+    DCHECK(web_request_api);
+    web_request_api->ProxyWebSocket(frame, std::move(factory), url,
+                                    site_for_cookies, user_agent,
+                                    std::move(handshake_client));
 #endif
+  } else {
+    proxy->Start(std::move(handshake_client));
+  }
 }
 
 void ChromeContentBrowserClient::WillCreateWebTransport(
--- a/chrome/browser/chrome_content_browser_client_receiver_bindings.cc
+++ b/chrome/browser/chrome_content_browser_client_receiver_bindings.cc
@@ -123,6 +123,18 @@
 #include "chrome/browser/offline_pages/offline_page_tab_helper.h"
 #endif
 
+// chromium-adblock
+#include "chrome/browser/cosmetic_filters/cosmetic_filters_tab_helper.h"
+#include "components/brave_shields/content/browser/ad_block_engine_wrapper.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
+#include "components/brave_shields/content/browser/domain_block_navigation_throttle.h"
+#include "components/brave_shields/core/common/brave_shield_constants.h"
+#include "components/brave_shields/core/common/features.h"
+#include "components/brave_shields/core/common/shields_settings.mojom.h"
+#include "components/cosmetic_filters/browser/cosmetic_filters_resources.h"
+#include "components/cosmetic_filters/common/cosmetic_filters.mojom.h"
+#include "mojo/public/cpp/bindings/self_owned_receiver.h"
+
 namespace {
 
 #if BUILDFLAG(SAFE_BROWSING_AVAILABLE)
@@ -203,6 +215,24 @@ void BindBadgeServiceForServiceWorker(
 }
 #endif
 
+void BindCosmeticFiltersResourcesOnTaskRunner(
+    mojo::PendingReceiver<cosmetic_filters::mojom::CosmeticFiltersResources>
+        receiver,
+    brave_shields::AdBlockEngineWrapper* engine_wrapper) {
+  mojo::MakeSelfOwnedReceiver(
+      std::make_unique<cosmetic_filters::CosmeticFiltersResources>(
+          engine_wrapper),
+      std::move(receiver));
+}
+
+void BindCosmeticFiltersResources(
+    content::RenderFrameHost* const frame_host,
+    mojo::PendingReceiver<cosmetic_filters::mojom::CosmeticFiltersResources>
+        receiver) {
+  g_browser_process->ad_block_service()->AsyncCall(base::BindOnce(
+      &BindCosmeticFiltersResourcesOnTaskRunner, std::move(receiver)));
+}
+
 }  // namespace
 
 void ChromeContentBrowserClient::ExposeInterfacesToRenderer(
@@ -317,6 +347,12 @@ void ChromeContentBrowserClient::BindMed
 void ChromeContentBrowserClient::RegisterBrowserInterfaceBindersForFrame(
     content::RenderFrameHost* render_frame_host,
     mojo::BinderMapWithContext<content::RenderFrameHost*>* map) {
+  // NOTE: This is weird. brave registers this after calling this function,
+  // but doing this at the end instead means that something tries to message
+  // this before it's bound.
+  map->Add<cosmetic_filters::mojom::CosmeticFiltersResources>(
+      base::BindRepeating(&BindCosmeticFiltersResources));
+
   chrome::internal::PopulateChromeFrameBinders(map, render_frame_host);
   chrome::internal::PopulateChromeWebUIFrameBinders(map, render_frame_host);
 
@@ -393,6 +429,11 @@ void ChromeContentBrowserClient::
     RegisterAssociatedInterfaceBindersForRenderFrameHost(
         content::RenderFrameHost& render_frame_host,
         blink::AssociatedInterfaceRegistry& associated_registry) {
+  associated_registry.AddInterface<
+      cosmetic_filters::mojom::CosmeticFiltersHandler>(base::BindRepeating(
+      &cosmetic_filters::CosmeticFiltersTabHelper::BindCosmeticFiltersHandler,
+      &render_frame_host));
+
   for (auto& ep : extra_parts_) {
     ep->ExposeInterfacesToRendererForRenderFrameHost(render_frame_host,
                                                      associated_registry);
--- a/chrome/browser/download/background_download_service_factory.cc
+++ b/chrome/browser/download/background_download_service_factory.cc
@@ -53,8 +53,27 @@
 #include "chrome/browser/ash/plugin_vm/plugin_vm_image_download_client.h"
 #endif
 
+// chromium-adblock
+#include "base/functional/bind.h"
+#include "chrome/browser/browser_process.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
+#include "components/brave_shields/content/browser/ad_block_subscription_download_client.h"
+#include "components/brave_shields/content/browser/ad_block_subscription_service_manager.h"
+#include "components/download/public/task/task_scheduler.h"
+
 namespace {
 
+brave_shields::AdBlockSubscriptionServiceManager*
+GetAdBlockSubscriptionServiceManager() {
+  return g_browser_process->ad_block_service()->subscription_service_manager();
+}
+
+std::unique_ptr<download::Client> CreateAdBlockSubscriptionDownloadClient(
+    Profile* profile) {
+  return std::make_unique<brave_shields::AdBlockSubscriptionDownloadClient>(
+      base::BindRepeating(&GetAdBlockSubscriptionServiceManager));
+}
+
 std::unique_ptr<download::Client> CreateBackgroundFetchDownloadClient(
     Profile* profile) {
   return std::make_unique<background_fetch::DownloadClient>(profile);
@@ -166,6 +185,11 @@ BackgroundDownloadServiceFactory::BuildS
   ProfileKey* profile_key = ProfileKey::FromSimpleFactoryKey(key);
 
   clients->insert(std::make_pair(
+      download::DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS,
+      std::make_unique<download::DeferredClientWrapper>(
+          base::BindOnce(&CreateAdBlockSubscriptionDownloadClient), key)));
+
+  clients->insert(std::make_pair(
       download::DownloadClient::BACKGROUND_FETCH,
       std::make_unique<download::DeferredClientWrapper>(
           base::BindOnce(&CreateBackgroundFetchDownloadClient), key)));
--- a/chrome/browser/net/BUILD.gn
+++ b/chrome/browser/net/BUILD.gn
@@ -53,6 +53,15 @@ source_set("net") {
     "secure_dns_util.h",
     "stub_resolver_config_reader.h",
     "system_network_context_manager.h",
+
+    # chromium-adblock
+    "brave_ad_block_csp_network_delegate_helper.h",
+    "brave_ad_block_tp_network_delegate_helper.h",
+    "brave_proxying_url_loader_factory.h",
+    "brave_proxying_web_socket.h",
+    "brave_request_handler.h",
+    "resource_context_data.h",
+    "url_context.h",
   ]
 
   public_deps = [
@@ -144,6 +153,15 @@ source_set("impl") {
     "secure_dns_util.cc",
     "stub_resolver_config_reader.cc",
     "system_network_context_manager.cc",
+
+    # chromium-adblock
+    "brave_ad_block_csp_network_delegate_helper.cc",
+    "brave_ad_block_tp_network_delegate_helper.cc",
+    "brave_proxying_url_loader_factory.cc",
+    "brave_proxying_web_socket.cc",
+    "brave_request_handler.cc",
+    "resource_context_data.cc",
+    "url_context.cc",
   ]
 
   public_deps = [
@@ -202,6 +220,11 @@ source_set("impl") {
     "//sandbox/policy",
     "//services/network:network_service",
     "//services/strings",
+
+    # chromium-adblock
+    "//components/brave_shields/content/browser",
+    "//components/brave_shields/core/browser",
+    "//components/brave_shields/core/common",
   ]
 
   if (is_chromeos) {
--- a/chrome/browser/prefs/browser_prefs.cc
+++ b/chrome/browser/prefs/browser_prefs.cc
@@ -106,6 +106,7 @@
 #include "components/accessibility_annotator/core/prefs.h"
 #include "components/autofill/core/common/autofill_prefs.h"
 #include "components/blocked_content/safe_browsing_triggered_popup_blocker.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
 #include "components/breadcrumbs/core/breadcrumbs_status.h"
 #include "components/browsing_data/core/pref_names.h"
 #include "components/certificate_transparency/pref_names.h"
@@ -1416,6 +1417,7 @@ void RegisterLocalState(PrefRegistrySimp
       registry);
 #endif
   autofill::prefs::RegisterLocalStatePrefs(registry);
+  brave_shields::RegisterPrefsForAdBlockService(registry);
   breadcrumbs::RegisterPrefs(registry);
   browser_shutdown::RegisterPrefs(registry);
   BrowserProcessImpl::RegisterPrefs(registry);
--- a/chrome/browser/profiles/chrome_browser_main_extra_parts_profiles.cc
+++ b/chrome/browser/profiles/chrome_browser_main_extra_parts_profiles.cc
@@ -55,6 +55,7 @@
 #include "chrome/browser/bitmap_fetcher/bitmap_fetcher_service_factory.h"
 #include "chrome/browser/bluetooth/bluetooth_chooser_context_factory.h"
 #include "chrome/browser/bookmarks/bookmark_model_factory.h"
+#include "chrome/browser/brave_shields/ad_block_pref_service_factory.h"
 #include "chrome/browser/breadcrumbs/breadcrumb_manager_keyed_service_factory.h"
 #include "chrome/browser/browser_process.h"
 #include "chrome/browser/browsing_data/browsing_data_history_observer_service.h"
@@ -1512,6 +1513,8 @@ void ChromeBrowserMainExtraPartsProfiles
   autofill::AtMemoryPromoTrackerFactory::GetInstance();
 #endif
 
+  brave_shields::AdBlockPrefServiceFactory::GetInstance();
+
   WebDataServiceFactory::GetInstance();
   webrtc_event_logging::WebRtcEventLogManagerKeyedServiceFactory::GetInstance();
 }
--- a/chrome/browser/renderer_context_menu/render_view_context_menu.cc
+++ b/chrome/browser/renderer_context_menu/render_view_context_menu.cc
@@ -362,6 +362,10 @@
 #include "ui/base/page_transition_types.h"
 #endif
 
+// chromium-adblock
+#include "chrome/browser/cosmetic_filters/cosmetic_filters_tab_helper.h"
+#include "components/brave_shields/core/common/features.h"
+
 using base::UserMetricsAction;
 using blink::ContextMenuData;
 using blink::ContextMenuDataEditFlags;
@@ -602,13 +606,14 @@ const std::map<int, int>& GetIdcToUmaMap
        {IDC_CONTENT_CONTEXT_GLIC, 162},
        {IDC_CONTENT_CONTEXT_VIDEO_FRAME, 163},
        {IDC_CONTENT_CONTEXT_LISTEN_TO_THIS_PAGE, 164},
+       {IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS, 165},
        // To add new items:
        //   - Add one more line above this comment block, using the UMA value
        //     from the line below this comment block.
        //   - Increment the UMA value in that latter line.
        //   - Add the new item to the RenderViewContextMenuItem enum in
        //     tools/metrics/histograms/metadata/ui/enums.xml.
-       {0, 165}});
+       {0, 166}});
   // LINT.ThenChange(//tools/metrics/histograms/metadata/ui/enums.xml:RenderViewContextMenuItem)
 
   // LINT.IfChange(ContextMenuOptionDesktop)
@@ -1655,13 +1660,29 @@ void RenderViewContextMenu::AppendDevelo
   bool hide_developer_items = IsDevToolsURL(params_.page_url) &&
                               params_.page_url.GetQuery().find(
                                   "debugFrontend=true") == std::string::npos;
-  if (hide_developer_items) {
+
+  bool add_block_elements =
+      (params_.selection_text.empty() || !params_.link_url.is_empty()) &&
+      params_.page_url.SchemeIsHTTPOrHTTPS() &&
+      base::FeatureList::IsEnabled(
+          brave_shields::features::kBraveShieldsElementPicker);
+
+  if (hide_developer_items && !add_block_elements) {
     return;
   }
 
   // In the DevTools popup menu, "developer items" is normally the only
   // section, so omit the separator there.
   menu_model_.AddSeparator(ui::NORMAL_SEPARATOR);
+
+  if (add_block_elements) {
+    menu_model_.AddItemWithStringId(IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS,
+                                    IDS_ADBLOCK_CONTEXT_BLOCK_ELEMENTS);
+  }
+
+  if (hide_developer_items)
+    return;
+
   if (content_type_->SupportsGroup(ContextMenuContentType::ITEM_GROUP_PAGE)) {
     menu_model_.AddItemWithStringId(IDC_VIEW_SOURCE,
                                     IDS_CONTENT_CONTEXT_VIEWPAGESOURCE);
@@ -2968,6 +2989,9 @@ bool RenderViewContextMenu::IsCommandIdE
   }
 #endif
   switch (id) {
+    case IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS:
+      return true;
+
     case IDC_BACK:
       return embedder_web_contents_->GetController().CanGoBack();
 
@@ -3287,6 +3311,11 @@ void RenderViewContextMenu::ExecuteComma
   }
 
   switch (id) {
+    case IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS:
+      cosmetic_filters::CosmeticFiltersTabHelper::LaunchContentPicker(
+          source_web_contents_);
+      break;
+
     case IDC_CONTENT_CONTEXT_OPENLINKNEWTAB: {
       WindowOpenDisposition new_tab_disposition =
           WindowOpenDisposition::NEW_BACKGROUND_TAB;
--- a/chrome/browser/resources/BUILD.gn
+++ b/chrome/browser/resources/BUILD.gn
@@ -19,6 +19,7 @@ assert(!is_fuchsia, "Fuchsia shouldn't u
 
 group("resources") {
   public_deps = [
+    "adblock_ui:resources",
     "contextual_tasks:resources",
     "glic:resources",
     "saved_tab_groups_unsupported:resources",
--- a/chrome/browser/ui/BUILD.gn
+++ b/chrome/browser/ui/BUILD.gn
@@ -76,6 +76,8 @@ static_library("ui") {
     "tab_ui_helper.h",
     "uninstall_browser_prompt.h",
     "view_ids.h",
+    "webui/brave_adblock_ui.cc",
+    "webui/brave_adblock_ui.h",
     "webui/browsing_topics/browsing_topics_internals_page_handler.cc",
     "webui/browsing_topics/browsing_topics_internals_page_handler.h",
     "webui/browsing_topics/browsing_topics_internals_ui.cc",
--- a/chrome/browser/ui/webui/chrome_web_ui_configs.cc
+++ b/chrome/browser/ui/webui/chrome_web_ui_configs.cc
@@ -92,6 +92,7 @@
 #include "chrome/browser/ui/webui/app_service_internals/app_service_internals_ui.h"
 #include "chrome/browser/ui/webui/autofill_ml_internals/autofill_ml_internals_ui.h"
 #include "chrome/browser/ui/webui/bookmarks/bookmarks_ui.h"
+#include "chrome/browser/ui/webui/brave_adblock_ui.h"
 #include "chrome/browser/ui/webui/color_pipeline_internals/color_pipeline_internals_ui.h"
 #include "chrome/browser/ui/webui/commerce/shopping_insights_side_panel_ui.h"
 #include "chrome/browser/ui/webui/downloads/downloads_ui.h"
@@ -247,6 +248,7 @@ void RegisterChromeWebUIConfigs() {
       std::make_unique<AccessibilityAnnotatorInternalsUIConfig>());
   map.AddWebUIConfig(std::make_unique<AutofillInternalsUIConfig>());
   map.AddWebUIConfig(std::make_unique<BluetoothInternalsUIConfig>());
+  map.AddWebUIConfig(std::make_unique<BraveAdblockUIConfig>());
   map.AddWebUIConfig(std::make_unique<BrowsingTopicsInternalsUIConfig>());
   map.AddWebUIConfig(
       std::make_unique<chrome_finds_internals::ChromeFindsInternalsUIConfig>());
--- a/chrome/chrome_paks.gni
+++ b/chrome/chrome_paks.gni
@@ -119,6 +119,7 @@ template("chrome_extra_paks") {
     output = "${invoker.output_dir}/resources.pak"
     sources = [
       "$root_gen_dir/base/tracing/protos/tracing_proto_resources.pak",
+      "$root_gen_dir/chrome/adblock_ui_resources.pak",
       "$root_gen_dir/chrome/browser/glic/resources/glic_browser_resources.pak",
       "$root_gen_dir/chrome/browser_resources.pak",
       "$root_gen_dir/chrome/chrome_unscaled_resources.pak",
@@ -175,6 +176,7 @@ template("chrome_extra_paks") {
       "//chrome/browser:resources",
       "//chrome/browser/glic/resources:browser_resources",
       "//chrome/browser/resources",
+      "//chrome/browser/resources/adblock_ui:resources",
       "//chrome/browser/resources/contextual_tasks:resources",
       "//chrome/browser/resources/glic:resources",
       "//chrome/browser/resources/saved_tab_groups_unsupported:resources",
--- a/chrome/common/webui_url_constants.cc
+++ b/chrome/common/webui_url_constants.cc
@@ -199,6 +199,7 @@ base::span<const base::cstring_view> Chr
 #endif
       content::kChromeUIPrivateAggregationInternalsHost,
       content::kChromeUIAttributionInternalsHost,
+      kAdblockHost,
       content::kChromeUIBlobInternalsHost,
       content::kChromeUIDinoHost,
       content::kChromeUIGpuHost,
--- a/chrome/common/webui_url_constants.h
+++ b/chrome/common/webui_url_constants.h
@@ -33,6 +33,7 @@ namespace chrome {
 // Not all components have corresponding URLs and vice versa. Only add as
 // needed.
 // Please keep in alphabetical order, with OS/feature specific sections below.
+inline constexpr char kAdblockHost[] = "adblock";
 inline constexpr char kChromeUIAboutHost[] = "about";
 inline constexpr char kChromeUIAboutURL[] = "chrome://about/";
 inline constexpr char kChromeUIAccessCodeCastHost[] = "access-code-cast";
--- a/chrome/renderer/BUILD.gn
+++ b/chrome/renderer/BUILD.gn
@@ -238,6 +238,10 @@ static_library("renderer") {
     "//third_party/widevine/cdm:buildflags",
     "//ui/gfx/geometry",
     "//v8",
+
+    # chromium-adblock
+    "//components/brave_shields/core/common",
+    "//components/cosmetic_filters/renderer",
   ]
 
   if (is_mac || is_win || is_linux || is_chromeos) {
--- a/chrome/renderer/chrome_content_renderer_client.cc
+++ b/chrome/renderer/chrome_content_renderer_client.cc
@@ -266,6 +266,10 @@
 #include "chrome/renderer/media/chrome_key_systems.h"
 #endif
 
+// chromium-adblock
+#include "components/brave_shields/core/common/features.h"
+#include "components/cosmetic_filters/renderer/cosmetic_filters_js_render_frame_observer.h"
+
 using autofill::AutofillAgent;
 using autofill::PasswordAutofillAgent;
 using autofill::PasswordGenerationAgent;
@@ -793,6 +797,15 @@ void ChromeContentRendererClient::Render
     WebUIBrowserRendererExtension::Create(render_frame);
   }
 #endif
+
+  if (base::FeatureList::IsEnabled(
+          brave_shields::features::kBraveAdblockCosmeticFiltering)) {
+    auto dynamic_params_closure = base::BindRepeating([]() { return false; });
+
+    new cosmetic_filters::CosmeticFiltersJsRenderFrameObserver(
+        render_frame, ISOLATED_WORLD_ID_CHROME_INTERNAL,
+        dynamic_params_closure);
+  }
 }
 
 void ChromeContentRendererClient::WebViewCreated(
@@ -1397,6 +1410,13 @@ ChromeContentRendererClient::CreateSpeec
 
 void ChromeContentRendererClient::RunScriptsAtDocumentStart(
     content::RenderFrame* render_frame) {
+  auto* observer =
+      cosmetic_filters::CosmeticFiltersJsRenderFrameObserver::Get(render_frame);
+  // Run this before any extensions
+  if (observer) {
+    observer->RunScriptsAtDocumentStart();
+  }
+
 #if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
   extensions::ExtensionsRendererClient::Get()->RunScriptsAtDocumentStart(
       render_frame);
--- a/components/component_updater/BUILD.gn
+++ b/components/component_updater/BUILD.gn
@@ -17,6 +17,8 @@ component("component_updater_paths") {
 
 static_library("component_updater") {
   sources = [
+    "browser/dat_file_util.cc",
+    "browser/dat_file_util.h",
     "component_installer.cc",
     "component_installer.h",
     "component_updater_command_line_config_policy.cc",
--- a/components/components_strings.grd
+++ b/components/components_strings.grd
@@ -295,6 +295,8 @@
       <part file="crash_strings.grdp" />
       <part file="dialog_strings.grdp" />
       <part file="dom_distiller_strings.grdp" />
+      <part file="domain_block_strings.grdp" />
+      <part file="element_picker_strings.grdp" />
       <part file="enterprise_strings.grdp" />
       <part file="error_page_strings.grdp" />
       <part file="find_in_page_strings.grdp" />
--- a/components/components_strings.grd.gritdeps
+++ b/components/components_strings.grd.gritdeps
@@ -17,6 +17,8 @@ components/contextual_tasks_strings.grdp
 components/crash_strings.grdp
 components/dialog_strings.grdp
 components/dom_distiller_strings.grdp
+components/domain_block_strings.grdp
+components/element_picker_strings.grdp
 components/enterprise_strings.grdp
 components/error_page_strings.grdp
 components/external_intents_strings.grdp
--- a/components/download/internal/background_service/proto/entry.proto
+++ b/components/download/internal/background_service/proto/entry.proto
@@ -28,6 +28,7 @@ enum DownloadClient {
 
   // New clients should be added above here.
   BOUNDARY = 11;
+  CUSTOM_LIST_SUBSCRIPTIONS = 256;
 }
 
 // Custom key value pair provided by the client and will be sent back to client
--- a/components/download/internal/background_service/proto_conversions.cc
+++ b/components/download/internal/background_service/proto_conversions.cc
@@ -61,6 +61,8 @@ protodb::DownloadClient ProtoConversions
       return protodb::DownloadClient::TEST_2;
     case DownloadClient::TEST_3:
       return protodb::DownloadClient::TEST_3;
+    case DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS:
+      return protodb::DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS;
     case DownloadClient::OFFLINE_PAGE_PREFETCH:
       return protodb::DownloadClient::OFFLINE_PAGE_PREFETCH;
     case DownloadClient::BACKGROUND_FETCH:
@@ -93,6 +95,8 @@ DownloadClient ProtoConversions::Downloa
       return DownloadClient::TEST_2;
     case protodb::DownloadClient::TEST_3:
       return DownloadClient::TEST_3;
+    case protodb::DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS:
+      return DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS;
     case protodb::DownloadClient::OFFLINE_PAGE_PREFETCH:
       return DownloadClient::OFFLINE_PAGE_PREFETCH;
     case protodb::DownloadClient::BACKGROUND_FETCH:
--- a/components/download/public/background_service/clients.cc
+++ b/components/download/public/background_service/clients.cc
@@ -14,6 +14,8 @@ std::string BackgroundDownloadClientToSt
     case DownloadClient::TEST_3:
     case DownloadClient::INVALID:
       return "__Test__";
+    case DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS:
+      return "CustomListSubscriptions";
     case DownloadClient::OFFLINE_PAGE_PREFETCH:
       return "OfflinePage";
     case DownloadClient::BACKGROUND_FETCH:
--- a/components/download/public/background_service/clients.h
+++ b/components/download/public/background_service/clients.h
@@ -32,6 +32,8 @@ enum class DownloadClient {
   // Represents an uninitialized DownloadClient variable.
   INVALID = 0,
 
+  CUSTOM_LIST_SUBSCRIPTIONS = 255,
+
   OFFLINE_PAGE_PREFETCH = 1,
 
   BACKGROUND_FETCH = 2,
--- a/components/resources/BUILD.gn
+++ b/components/resources/BUILD.gn
@@ -86,6 +86,8 @@ grit("components_resources") {
 
   deps = [
     ":about_credits",
+    "//components/brave_shields/resources:bundle_js",
+    "//components/cosmetic_filters/resources",
     "//components/dom_distiller/core:dom_distiller_viewer_js",
     "//components/neterror/resources:bundle_js",
     "//components/security_interstitials/content/resources:build_ts",
--- a/components/resources/components_resources.grd
+++ b/components/resources/components_resources.grd
@@ -15,6 +15,8 @@ dev_ui_components_resources.grd.
     <includes>
       <part file="about_ui_resources.grdp" />
       <part file="commerce_resources.grdp" />
+      <part file="cosmetic_filters_resources.grdp" />
+      <part file="domain_block_resources.grdp" />
       <part file="dom_distiller_resources.grdp" />
       <part file="neterror_resources.grdp" />
       <part file="notifications_global_cache_list.grdp" />
--- a/components/value_store/value_store_frontend.cc
+++ b/components/value_store/value_store_frontend.cc
@@ -33,6 +33,37 @@ class ValueStoreFrontend::Backend : publ
   Backend(const Backend&) = delete;
   Backend& operator=(const Backend&) = delete;
 
+  void GetKeys(ValueStoreFrontend::ReadCallback callback) {
+    DCHECK(file_task_runner_->RunsTasksInCurrentSequence());
+    LazyInit();
+    ValueStore::ReadResult result = storage_->GetKeys();
+
+    std::optional<base::Value> value;
+    if (result.status().ok()) {
+      value = base::Value(base::Value::Type::DICT);
+      base::DictValue& dict_value = value->GetDict();
+      for (auto [key, _] : result.settings()) {
+        // NOTE: GetKeys() populates the values with dummy values.
+        // We have to call Get on the key itself to retrieve
+        // the actual values.
+        ValueStore::ReadResult value_result = storage_->Get(key);
+        if (!value_result.status().ok()) {
+            LOG(WARNING) << "Reading " << key << " from " << db_path_.value()
+                         << " failed: " << result.status().message;
+            continue;
+        }
+        dict_value.Set(key, *value_result.settings().Extract(key));
+      }
+    } else {
+      LOG(WARNING) << "Reading keys from " << db_path_.value()
+                   << " failed: " << result.status().message;
+    }
+
+    origin_task_runner_->PostTask(
+        FROM_HERE, base::BindOnce(&ValueStoreFrontend::Backend::RunCallback,
+                                  this, std::move(callback), std::move(value)));
+  }
+
   void Get(const std::string& key, ValueStoreFrontend::ReadCallback callback) {
     DCHECK(file_task_runner_->RunsTasksInCurrentSequence());
     LazyInit();
@@ -127,6 +158,14 @@ ValueStoreFrontend::~ValueStoreFrontend(
   DCHECK(origin_task_runner_->RunsTasksInCurrentSequence());
 }
 
+void ValueStoreFrontend::GetKeys(ReadCallback callback) {
+  DCHECK(origin_task_runner_->RunsTasksInCurrentSequence());
+
+  file_task_runner_->PostTask(
+      FROM_HERE, base::BindOnce(&ValueStoreFrontend::Backend::GetKeys, backend_,
+                                std::move(callback)));
+}
+
 void ValueStoreFrontend::Get(const std::string& key, ReadCallback callback) {
   DCHECK(origin_task_runner_->RunsTasksInCurrentSequence());
 
--- a/components/value_store/value_store_frontend.h
+++ b/components/value_store/value_store_frontend.h
@@ -35,6 +35,8 @@ class ValueStoreFrontend {
   ValueStoreFrontend(const ValueStoreFrontend&) = delete;
   ValueStoreFrontend& operator=(const ValueStoreFrontend&) = delete;
 
+  void GetKeys(ValueStoreFrontend::ReadCallback callback);
+
   // Retrieves a value from the database asynchronously, passing a copy to
   // |callback| when ready. NULL is passed if no matching entry is found.
   void Get(const std::string& key, ReadCallback callback);
--- a/content/browser/BUILD.gn
+++ b/content/browser/BUILD.gn
@@ -874,6 +874,8 @@ source_set("browser") {
     "devtools/cross_thread_protocol_callback.h",
     "devtools/dedicated_worker_devtools_agent_host.cc",
     "devtools/dedicated_worker_devtools_agent_host.h",
+    "devtools/adblock_devtools_instrumentation.cc",
+    "devtools/adblock_devtools_instrumentation.h",
     "devtools/devtools_agent_host_impl.cc",
     "devtools/devtools_agent_host_impl.h",
     "devtools/devtools_background_services_context_impl.cc",
--- a/content/browser/devtools/protocol/network_handler.cc
+++ b/content/browser/devtools/protocol/network_handler.cc
@@ -1576,6 +1576,16 @@ std::vector<NetworkHandler*> NetworkHand
   return host->HandlersByName<NetworkHandler>(Network::Metainfo::domainName);
 }
 
+void NetworkHandler::RequestAdblockInfoReceived(
+    const std::string& request_id,
+    std::unique_ptr<protocol::Network::AdblockInfo> info) {
+  if (!enabled_) {
+    return;
+  }
+
+  frontend_->RequestAdblockInfoReceived(request_id, std::move(info));
+}
+
 void NetworkHandler::Wire(UberDispatcher* dispatcher) {
   frontend_ = std::make_unique<Network::Frontend>(dispatcher->channel());
   Network::Dispatcher::wire(dispatcher, this);
--- a/content/browser/devtools/protocol/network_handler.h
+++ b/content/browser/devtools/protocol/network_handler.h
@@ -297,6 +297,11 @@ class NetworkHandler : public DevToolsDo
 
   void NavigationRequestWillBeSent(const NavigationRequest& nav_request,
                                    base::TimeTicks timestamp);
+
+  void RequestAdblockInfoReceived(
+      const std::string& request_id,
+      std::unique_ptr<protocol::Network::AdblockInfo> info);
+
   void FencedFrameReportRequestSent(const std::string& request_id,
                                     const network::ResourceRequest& request,
                                     const std::string& event_data,
--- a/content/browser/devtools/protocol_config.json
+++ b/content/browser/devtools/protocol_config.json
@@ -134,7 +134,7 @@
                   "setCookieControls",
                   "configureDurableMessages"
                 ],
-                "include_events": ["requestWillBeSent", "responseReceived", "loadingFinished", "loadingFailed", "requestIntercepted", "signedExchangeReceived", "requestWillBeSentExtraInfo", "responseReceivedExtraInfo", "responseReceivedEarlyHints", "trustTokenOperationDone", "policyUpdated", "reportingApiReportAdded", "reportingApiReportUpdated", "reportingApiEndpointsChangedForOrigin", "deviceBoundSessionsAdded", "deviceBoundSessionEventOccurred"],
+                "include_events": ["requestAdblockInfoReceived", "requestWillBeSent", "responseReceived", "loadingFinished", "loadingFailed", "requestIntercepted", "signedExchangeReceived", "requestWillBeSentExtraInfo", "responseReceivedExtraInfo", "responseReceivedEarlyHints", "trustTokenOperationDone", "policyUpdated", "reportingApiReportAdded", "reportingApiReportUpdated", "reportingApiEndpointsChangedForOrigin", "deviceBoundSessionsAdded", "deviceBoundSessionEventOccurred"],
                 "async": [
                   "configureDurableMessages",
                   "clearBrowserCookies",
--- a/content/public/browser/BUILD.gn
+++ b/content/public/browser/BUILD.gn
@@ -96,6 +96,7 @@ source_set("browser_sources") {
     "cdm_registry.h",
     "cdm_storage_data_model.h",
     "certificate_request_result_type.h",
+    "//content/public/browser/devtools/adblock_devtools_instumentation.h",
     "child_process_data.cc",
     "child_process_data.h",
     "child_process_host.h",
--- a/third_party/blink/public/devtools_protocol/domains/Network.pdl
+++ b/third_party/blink/public/devtools_protocol/domains/Network.pdl
@@ -11,6 +11,24 @@ domain Network
   depends on Runtime
   depends on Security
 
+  experimental type AdblockInfo extends object
+    properties
+      string requestUrl
+      string checkedUrl
+      string sourceHost
+      optional ResourceType resourceType
+      boolean blocked
+      boolean didMatchImportantRule
+      boolean didMatchRule
+      boolean didMatchException
+      boolean hasMockData
+      optional string rewrittenUrl
+
+  experimental event requestAdblockInfoReceived
+    parameters
+      Network.RequestId requestId
+      AdblockInfo info
+
   # Resource type as it was perceived by the rendering engine.
   type ResourceType extends string
     enum
--- a/third_party/devtools-frontend/src/config/gni/devtools_grd_files.gni
+++ b/third_party/devtools-frontend/src/config/gni/devtools_grd_files.gni
@@ -1811,6 +1811,7 @@ grd_files_unbundled_sources = [
   "front_end/panels/network/NetworkPanel.js",
   "front_end/panels/network/NetworkSearchScope.js",
   "front_end/panels/network/NetworkWaterfallColumn.js",
+  "front_end/panels/network/RequestAdblockView.js",
   "front_end/panels/network/RequestConditionsDrawer.js",
   "front_end/panels/network/RequestCookiesView.js",
   "front_end/panels/network/RequestDeviceBoundSessionsView.js",
--- a/third_party/devtools-frontend/src/front_end/core/protocol_client/InspectorBackend.ts
+++ b/third_party/devtools-frontend/src/front_end/core/protocol_client/InspectorBackend.ts
@@ -77,6 +77,7 @@ export class InspectorBackend implements
     globalThis.Protocol ||= {};
 
     InspectorBackendCommands.registerCommands(this);
+    this.registerEvent('Network.requestAdblockInfoReceived' as QualifiedName, ['requestId', 'info']);
   }
 
   private getOrCreateEventParameterNamesForDomain(domain: ProtocolDomainName): EventParameterNames {
--- a/third_party/devtools-frontend/src/front_end/core/sdk/NetworkManager.ts
+++ b/third_party/devtools-frontend/src/front_end/core/sdk/NetworkManager.ts
@@ -630,6 +630,8 @@ export class NetworkDispatcher implement
     }
   }
 
+  #requestsAdblockInfo: Map<string, object> | undefined
+
   private headersMapToHeadersArray(headersMap: Protocol.Network.Headers): NameValue[] {
     const result = [];
     for (const name in headersMap) {
@@ -747,6 +749,21 @@ export class NetworkDispatcher implement
     }
   }
 
+  adblockInfoForId(id: string): object | null {
+    return this.#requestsAdblockInfo?.get(id) || null
+  }
+
+  requestAdblockInfoReceived(params: any): void {
+    if (!this.#requestsAdblockInfo) {
+      this.#requestsAdblockInfo = new Map()
+    }
+    this.#requestsAdblockInfo.set(params.requestId, params.info)
+    const networkRequest = this.requestForId(params.requestId)
+    if (networkRequest) {
+      ;(this as any).updateNetworkRequest(networkRequest)
+    }
+  }
+
   requestForId(id: string): NetworkRequest|null {
     return this.#requestsById.get(id) || null;
   }
@@ -1267,6 +1284,15 @@ export class NetworkDispatcher implement
   }
 
   clearRequests(): void {
+    if (this.#requestsAdblockInfo) {
+      for (const [requestId, _] of this.#requestsAdblockInfo) {
+        const request = this.requestForId(requestId)
+        if (request?.finished) {
+          this.#requestsAdblockInfo.delete(requestId)
+        }
+      }
+    }
+
     for (const [requestId, request] of this.#requestsById) {
       if (request.finished) {
         this.#requestsById.delete(requestId);
--- a/third_party/devtools-frontend/src/front_end/panels/network/BUILD.gn
+++ b/third_party/devtools-frontend/src/front_end/panels/network/BUILD.gn
@@ -48,6 +48,7 @@ devtools_ui_module("network") {
     "NetworkPanel.ts",
     "NetworkSearchScope.ts",
     "NetworkWaterfallColumn.ts",
+    "RequestAdblockView.ts",
     "RequestConditionsDrawer.ts",
     "RequestCookiesView.ts",
     "RequestDeviceBoundSessionsView.ts",
--- a/third_party/devtools-frontend/src/front_end/panels/network/NetworkDataGridNode.ts
+++ b/third_party/devtools-frontend/src/front_end/panels/network/NetworkDataGridNode.ts
@@ -1223,6 +1223,26 @@ export class NetworkRequestNode extends
     this.parentView().dispatchEventToListeners(Events.RequestSelected, this.requestInternal);
   }
 
+  override updateBackgroundColor(): void {
+    const networkManager = SDK.NetworkManager.NetworkManager.forRequest(
+      this.request()
+    )
+    if (networkManager && networkManager.dispatcher) {
+      const info = (networkManager.dispatcher as any).adblockInfoForId(
+        this.request().requestId()
+      )
+      const element = this.existingElement() as HTMLElement
+      if (info && element) {
+        if ('blocked' in info && info.blocked) {
+          element.style.color = 'var(--sys-color-error)'
+        } else if ('didMatchException' in info && info.didMatchException) {
+          element.style.color = 'var(--ref-palette-green60)'
+        }
+      }
+    }
+    super.updateBackgroundColor()
+  }
+
   private openInNewTab(): void {
     Host.InspectorFrontendHost.InspectorFrontendHostInstance.openInNewTab(this.requestInternal.url());
   }
--- a/third_party/devtools-frontend/src/front_end/panels/network/NetworkItemView.ts
+++ b/third_party/devtools-frontend/src/front_end/panels/network/NetworkItemView.ts
@@ -6,6 +6,7 @@
 
 import * as Common from '../../core/common/common.js';
 import * as i18n from '../../core/i18n/i18n.js';
+import * as LegacyWrapper from '../../ui/components/legacy_wrapper/legacy_wrapper.js'
 import * as Platform from '../../core/platform/platform.js';
 import * as SDK from '../../core/sdk/sdk.js';
 import type * as NetworkTimeCalculator from '../../models/network_time_calculator/network_time_calculator.js';
@@ -17,6 +18,7 @@ import * as VisualLogging from '../../ui
 
 import * as NetworkComponents from './components/components.js';
 import {EventSourceMessagesView} from './EventSourceMessagesView.js';
+import {RequestAdblockView} from './RequestAdblockView.js';
 import {RequestCookiesView} from './RequestCookiesView.js';
 import {RequestDeviceBoundSessionsView} from './RequestDeviceBoundSessionsView.js';
 import {RequestHeadersView} from './RequestHeadersView.js';
@@ -255,6 +257,15 @@ export class NetworkItemView extends UI.
           trustTokensView, i18nString(UIStrings.trustTokenOperationDetails));
     }
 
+    this.appendTab(
+        'blocking',
+        'Adblock',
+        LegacyWrapper.LegacyWrapper.legacyWrapper(
+          UI.Widget.VBox,
+          new RequestAdblockView(this.request())
+        ),
+        'Adblock')
+
     this.#initialTab = initialTab || this.#resourceViewTabSetting.get();
     // Selecting tabs should not be handled by the super class.
     this.setAutoSelectFirstItemOnShow(false);
--- a/third_party/devtools-frontend/src/third_party/i18n/collect-strings.js
+++ b/third_party/devtools-frontend/src/third_party/i18n/collect-strings.js
@@ -27,6 +27,7 @@ export const IGNORED_PATH_COMPONENTS = [
   '**/*_test_runner/**',
   '**/third_party/**',
   '**/*.test.ts',
+  '**/RequestAdblockView.ts',
 ];
 
 /**
--- a/third_party/rust/chromium_crates_io/Cargo.toml
+++ b/third_party/rust/chromium_crates_io/Cargo.toml
@@ -59,6 +59,16 @@ skrifa = "0.42.1"
 small_ctor = "0.1"
 static_assertions = "1"
 xml = "1.2.0"
+#
+# chromium-adblock
+#
+lifeguard = "0.6"
+thiserror_V1 = { package = "thiserror", version = "1" }
+
+[dependencies.adblock]
+version = "0.12.5"
+default-features = false
+features = ["full-regex-handling", "debug-info", "css-validation"]
 
 [dependencies.bytemuck]
 version = "1"
--- a/third_party/rust/chromium_crates_io/gnrt_config.toml
+++ b/third_party/rust/chromium_crates_io/gnrt_config.toml
@@ -116,12 +116,10 @@ extra_kv = { allow_unsafe = true }
 allow_unsafe = false
 
 [crate.ahash]
-group = 'test'
 allow_first_party_usage = false
 extra_kv = { allow_unsafe = true }
 
 [crate.aho-corasick]
-group = 'test'
 allow_first_party_usage = false
 extra_kv = { allow_unsafe = true }
 
@@ -340,8 +338,8 @@ remove_deps = ['winapi']
 extra_kv = { allow_unsafe = true }
 
 [crate.getrandom]
-group = 'test'
 allow_first_party_usage = false
+extra_input_roots = ['../README.md']
 
 # TODO: Remove multiple versions of the `getrandom` crate:
 # getrandom@0.3.4, getrandom@0.4.1
@@ -788,12 +786,16 @@ allow_unsafe = false
 [crate.termcolor.extra_kv]
 allow_unsafe = false
 
-[crate.thiserror]
+[crate."thiserror@v2"]
 build_script_outputs = [ "private.rs" ]
 extra_kv = { allow_unsafe = false }
 
 [crate.thiserror-impl]
-extra_kv = { allow_unsafe = false }
+extra_kv = { raw_gn = '''
+if (epoch == "2") {
+  output_name = "thiserror_impl_v2"
+}
+''' , allow_unsafe = false }
 
 [crate.timezone_provider]
 extra_input_roots = ["**/*.data"]
@@ -895,6 +897,7 @@ allow_unsafe = false
 
 [crate.zerocopy]
 allow_first_party_usage = false
+extra_input_roots = ['../benches/transmute.x86-64']
 extra_kv = { allow_unsafe = true }
 
 [crate.zerofrom.extra_kv]
@@ -922,3 +925,125 @@ extra_kv = { allow_unsafe = true }
 [crate.zoneinfo64]
 extra_input_roots = ["data/zoneinfo64.res"]
 extra_kv = { allow_unsafe = true }
+
+#### chromium-adblock ####
+
+[crate.adblock]
+extra_kv = { raw_gn = '''
+features += [
+  "object-pooling",
+  "lifeguard",
+  "unsync-regex-caching",
+]
+deps += [ "//third_party/rust/lifeguard/v0_6:lib" ]
+''', allow_unsafe = true }
+group = 'safe'
+license_files = ['LICENSE']
+
+[crate.adblock-cxx]
+extra_kv = { allow_unsafe = false }
+
+[crate.addr]
+extra_kv = { allow_unsafe = false }
+
+[crate.cssparser]
+extra_kv = { allow_unsafe = true }
+
+[crate.cssparser-macros]
+extra_kv = { allow_unsafe = false }
+
+[crate.derive_more]
+extra_kv = { allow_unsafe = false }
+
+[crate.dtoa]
+extra_kv = { allow_unsafe = true }
+
+[crate.dtoa-short]
+extra_kv = { allow_unsafe = true }
+
+[crate.flatbuffers.extra_kv]
+allow_unsafe = true
+allow_unstable_features = ["trusted_len"]
+
+[crate.form_urlencoded]
+extra_kv = { allow_unsafe = true }
+
+[crate.fxhash]
+extra_kv = { allow_unsafe = false }
+
+[crate.idna]
+extra_kv = { allow_unsafe = true }
+
+[crate.idna_adapter]
+extra_kv = { allow_unsafe = false }
+
+[crate.lifeguard]
+extra_kv = { allow_unsafe = false }
+
+[crate.new_debug_unreachable]
+extra_kv = { allow_unsafe = false }
+
+[crate.percent-encoding]
+extra_kv = { allow_unsafe = true }
+
+[crate.phf]
+extra_kv = { allow_unsafe = false }
+
+[crate.phf_codegen]
+extra_kv = { allow_unsafe = false }
+
+[crate.phf_generator]
+extra_kv = { allow_unsafe = false }
+
+[crate.phf_macros]
+extra_kv = { allow_unsafe = false }
+
+[crate.phf_shared]
+extra_kv = { allow_unsafe = false }
+
+[crate.precomputed-hash]
+extra_kv = { allow_unsafe = false }
+
+[crate.psl]
+extra_kv = { allow_unsafe = false }
+
+[crate.psl-types]
+extra_kv = { allow_unsafe = false }
+
+[crate.rand]
+extra_kv = { allow_unsafe = true }
+
+[crate.rand_core]
+extra_kv = { allow_unsafe = true }
+
+[crate.regex]
+extra_kv = { allow_unsafe = true }
+
+[crate.rustc-hash]
+extra_kv = { allow_unsafe = false }
+
+[crate.rustc_version]
+extra_kv = { allow_unsafe = false }
+
+[crate.seahash]
+extra_kv = { allow_unsafe = true }
+
+[crate.selectors]
+build_script_outputs = [ "ascii_case_insensitive_html_attributes.rs" ]
+extra_kv = { allow_unsafe = true }
+
+[crate.semver]
+extra_kv = { allow_unsafe = true }
+
+[crate.servo_arc]
+extra_kv = { allow_unsafe = true }
+
+[crate.siphasher]
+extra_input_roots = ['../README.md']
+extra_kv = { allow_unsafe = true }
+
+[crate."thiserror@v1"]
+extra_kv = { allow_unsafe = false }
+
+[crate.url]
+extra_kv = { allow_unsafe = true }
--- a/tools/crates/gnrt/lib/readme.rs
+++ b/tools/crates/gnrt/lib/readme.rs
@@ -139,30 +139,6 @@ fn readme_file_from_package<'a>(
         Vec::new()
     };
 
-    if license_files.is_empty() {
-        bail!(
-            r#"License file not found for crate `{name}-{version}`.
-
-* If a license file exists but `gnrt` can't find it under
-  `{crate_vendor_dir}` then,
-    * Specify `license_files` in `[crate.{name}]`
-      section of `chromium_crates_io/gnrt_config.toml`
-    * Or tweak the `LICENSE_KIND_TO_LICENSE_FILES` map in
-      `tools/crates/gnrt/lib/readme.rs` to teach `gnrt`
-      about alternative filenames
-* If the crate didn't publish the license, then this may need
-  to be fixed upstream before the crate can be imported.
-  See also:
-    - https://crbug.com/369075726
-    - https://github.com/brendanzab/codespan/pull/355
-    - https://github.com/rust-lang/rustc-demangle/issues/72
-    - https://github.com/udoprog/relative-path/pull/60"#,
-            name = package.name(),
-            version = package.version(),
-            crate_vendor_dir = crate_vendor_dir.display(),
-        );
-    }
-
     let revision = {
         if let Ok(file) = std::fs::File::open(
             get_vendor_dir_for_package(paths, package.name(), package.version())
--- a/tools/crates/run_gnrt.py
+++ b/tools/crates/run_gnrt.py
@@ -23,6 +23,7 @@ from run_cargo import (RunCargo, DEFAULT
 def RunGnrt(rust_sysroot, out_dir, gnrt_args):
     target_dir = os.path.abspath(os.path.join(out_dir, 'target'))
     home_dir = os.path.abspath(os.path.join(target_dir, 'cargo_home'))
+    home_dir = None  # chromium-adblock
     cargo_args = [
         '--locked', 'run', '--release', '--manifest-path', GNRT_MANIFEST_PATH,
         '--target-dir', target_dir, '--'
--- a/tools/gritsettings/resource_ids.spec
+++ b/tools/gritsettings/resource_ids.spec
@@ -1250,6 +1250,10 @@
     "META": {"sizes": {"includes": [5]}},
     "includes": [7270],
   },
+  "<(SHARED_INTERMEDIATE_DIR)/chrome/browser/resources/adblock_ui/resources.grd": {
+    "META": {"sizes": {"includes": [2]}},
+    "includes": [7275],
+  },
   "<(SHARED_INTERMEDIATE_DIR)/chrome/browser/resources/autofill_ml_internals/resources.grd": {
     "META": {"sizes": {"includes": [20]}},
     "includes": [7280],
--- a/tools/metrics/histograms/metadata/ui/enums.xml
+++ b/tools/metrics/histograms/metadata/ui/enums.xml
@@ -363,6 +363,7 @@ chromium-metrics-reviews@google.com.
   <int value="162" label="IDC_CONTENT_CONTEXT_GLIC"/>
   <int value="163" label="IDC_CONTENT_CONTEXT_VIDEO_FRAME"/>
   <int value="164" label="IDC_CONTENT_CONTEXT_LISTEN_TO_THIS_PAGE"/>
+  <int value="165" label="IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS"/>
 </enum>
 
 <!-- LINT.ThenChange(//chrome/browser/renderer_context_menu/render_view_context_menu.cc:RenderViewContextMenuItem) -->
--- a/tools/typescript/validate_tsconfig.py
+++ b/tools/typescript/validate_tsconfig.py
@@ -134,6 +134,12 @@ def validateJavaScriptAllowed(source_dir
   if platform == 'ios' and '/ios/' in out_dir:
     return True, None
 
+  ui = 'chrome/browser/resources/adblock_ui'
+  if (source_dir.endswith(ui) or source_dir.endswith(ui + '/preprocessed')):
+    # I really didn't feel like using typescript for this, so it shall now no
+    # longer be a requirement :P
+    return True, None
+
   if platform == 'chromeos_ash':
     # Anything in these ChromeOS-specific directories is allowed to use allow_js.
     # TODO (rbpotter): If possible, standardize the build setup in some of these
--- a/ui/webui/resources/tools/build_webui.gni
+++ b/ui/webui/resources/tools/build_webui.gni
@@ -558,6 +558,7 @@ template("build_webui") {
         "//chrome/browser/resources/about_sys",
         "//chrome/browser/resources/access_code_cast",
         "//chrome/browser/resources/accessibility",
+        "//chrome/browser/resources/adblock_ui",
         "//chrome/browser/resources/bookmarks",
         "//chrome/browser/resources/browser_switch",
         "//chrome/browser/resources/certificate_manager",
